File: shapes_core.js
// 🛸 Shapes Core JS — Full API for Vanilla SPA
// Complete port of shapes_core.py

// --- Tokenizer ---
class ShapeTokenizer {
    constructor(strategy = 'part') {
        this.strategy = strategy;
        this.vocab = {};
        this.inverse_vocab = {};
        this.is_trained = false;
    }

    train(scenes) {
        if (this.strategy === 'pixel') this._train_pixel();
        else if (this.strategy === 'shape') this._train_shape(scenes);
        else this._train_part();
        this.is_trained = true;
    }

    _train_pixel() {
        const colors = ["empty", ...Object.keys(LAYER_COLORS)];
        colors.forEach((c, i) => {
            this.vocab[c] = i;
            this.inverse_vocab[i] = c;
        });
    }

    _train_shape(scenes) {
        this.vocab = { "<UNK>": 0 };
        let idx = 1;
        scenes.forEach(scene => {
            (scene.shapes || []).forEach(s => {
                const key = `${s.type}_${s.color}`;
                if (!(key in this.vocab)) {
                    this.vocab[key] = idx++;
                }
            });
        });
        Object.entries(this.vocab).forEach(([k, v]) => this.inverse_vocab[v] = k);
    }

    _train_part() {
        SHAPE_PARTS.forEach((p, i) => {
            this.vocab[p] = i;
            this.inverse_vocab[i] = p;
        });
    }

    encode(scene) {
        if (this.strategy === 'pixel') return this._encode_pixel(scene);
        if (this.strategy === 'shape') return this._encode_shape(scene);
        return this._encode_part(scene);
    }

    _encode_pixel(scene, gridSize = 10) {
        const grid = Array(gridSize).fill(0).map(() => Array(gridSize).fill(this.vocab["empty"]));
        (scene.shapes || []).forEach(s => {
            const color_id = this.vocab[s.color] !== undefined ? this.vocab[s.color] : this.vocab["empty"];
            const x0 = Math.floor(s.x), y0 = Math.floor(s.y);
            const x1 = Math.min(x0 + Math.ceil(s.w), gridSize);
            const y1 = Math.min(y0 + Math.ceil(s.h), gridSize);
            for (let r = Math.max(0, y0); r < y1; r++) {
                for (let c = Math.max(0, x0); c < x1; c++) {
                    grid[r][c] = color_id;
                }
            }
        });
        return grid.flat();
    }

    _encode_shape(scene) {
        const unk = this.vocab["<UNK>"] || 0;
        return (scene.shapes || []).map(s => {
            const key = `${s.type}_${s.color}`;
            return this.vocab[key] !== undefined ? this.vocab[key] : unk;
        });
    }

    _encode_part(scene) {
        const tokens = [];
        (scene.shapes || []).forEach(s => {
            const parts = SHAPE_DECOMPOSITION[s.type] || ["straight_edge"];
            parts.forEach(p => tokens.push(this.vocab[p] || 0));
        });
        return tokens;
    }

    decode(token_ids) {
        return token_ids.map(id => this.inverse_vocab[id] || "?");
    }

    get vocab_size() {
        return Object.keys(this.vocab).length;
    }
}

// --- Embeddings ---
class ShapeEmbedding {
    constructor(strategy = 'spatial', dimensions = 16) {
        this.strategy = strategy;
        this.dimensions = dimensions;
        this.scene_names = [];
        this.embeddings = [];
    }

    train(scenesDict) {
        this.scene_names = Object.keys(scenesDict);
        this.embeddings = this.scene_names.map(name => {
            return this._compute_embedding(scenesDict[name]);
        });
    }

    _compute_embedding(scene) {
        const shapes = scene.shapes || [];
        const dim = this.dimensions;
        const vec = new Array(dim).fill(0);

        if (shapes.length === 0) return vec;

        // Feature extraction: spread shape properties across dimensions
        shapes.forEach((s, idx) => {
            vec[idx % dim] += s.x / 10.0;
            vec[(idx + 1) % dim] += s.y / 10.0;
            vec[(idx + 2) % dim] += s.w / 5.0;
            vec[(idx + 3) % dim] += s.h / 5.0;
            // Color hash
            let colorHash = 0;
            for (let i = 0; i < s.color.length; i++) colorHash += s.color.charCodeAt(i);
            vec[(idx + 4) % dim] += (colorHash % 10) / 10.0;
            // Type hash
            let typeHash = 0;
            for (let i = 0; i < s.type.length; i++) typeHash += s.type.charCodeAt(i);
            vec[(idx + 5) % dim] += (typeHash % 10) / 10.0;
        });

        // Normalize
        let mag = 0;
        vec.forEach(v => mag += v * v);
        mag = Math.sqrt(mag);
        if (mag > 0) {
            for (let i = 0; i < dim; i++) vec[i] /= mag;
        }

        return vec;
    }

    get_embedding(scene) {
        // If scene is a string (name), look it up
        if (typeof scene === 'string') {
            const idx = this.scene_names.indexOf(scene);
            if (idx >= 0) return this.embeddings[idx];
            // Fallback: compute from SAMPLE_SCENES if available
            if (typeof SAMPLE_SCENES !== 'undefined' && SAMPLE_SCENES[scene]) {
                return this._compute_embedding(SAMPLE_SCENES[scene]);
            }
            return new Array(this.dimensions).fill(0);
        }
        // Otherwise it's a scene dict — compute directly
        return this._compute_embedding(scene);
    }
}

// --- Vector Database ---
class ShapeVectorDB {
    constructor(dimensions = 16) {
        this.dimensions = dimensions;
        this.ids = [];
        this.vectors = [];
    }

    add_batch(ids, vectors) {
        for (let i = 0; i < ids.length; i++) {
            this.ids.push(ids[i]);
            this.vectors.push(vectors[i]);
        }
    }

    search(queryVector, k = 3) {
        const results = [];
        for (let i = 0; i < this.ids.length; i++) {
            const sim = this._cosineSimilarity(queryVector, this.vectors[i]);
            results.push([this.ids[i], sim]);
        }
        results.sort((a, b) => b[1] - a[1]);
        return results.slice(0, k);
    }

    _cosineSimilarity(a, b) {
        let dot = 0, na = 0, nb = 0;
        for (let i = 0; i < a.length; i++) {
            dot += a[i] * b[i];
            na += a[i] * a[i];
            nb += b[i] * b[i];
        }
        if (na === 0 || nb === 0) return 0;
        return dot / (Math.sqrt(na) * Math.sqrt(nb));
    }
}

// --- Attention ---
class ShapeAttention {
    compute_attention(scene) {
        const shapes = scene.shapes || [];
        const n = shapes.length;
        const labels = shapes.map(s => s.label);
        
        if (n === 0) return [[], []];

        const proximity = this._proximity_scores(shapes);
        const color = this._color_scores(shapes);
        const alignment = this._alignment_scores(shapes);
        const containment = this._containment_scores(shapes);

        const weights = [proximity, color, alignment, containment].map(matrix => 
            matrix.map(row => this._softmax(row))
        );

        // Return as array tuple [weights, labels] to match app.js expectations
        return [weights, labels];
    }

    _softmax(arr) {
        const max = Math.max(...arr);
        const exps = arr.map(x => Math.exp(x - max));
        const sum = exps.reduce((a, b) => a + b, 0);
        return exps.map(x => x / sum);
    }

    _center(s) {
        return [s.x + s.w / 2, s.y + s.h / 2];
    }

    _proximity_scores(shapes) {
        const n = shapes.length;
        const scores = Array(n).fill().map(() => Array(n).fill(0));
        for (let i = 0; i < n; i++) {
            const [cx_i, cy_i] = this._center(shapes[i]);
            for (let j = 0; j < n; j++) {
                if (i === j) continue;
                const [cx_j, cy_j] = this._center(shapes[j]);
                const dist = Math.sqrt((cx_i - cx_j)**2 + (cy_i - cy_j)**2);
                scores[i][j] = Math.max(0, 10 - dist);
            }
        }
        return scores;
    }

    _color_scores(shapes) {
        const n = shapes.length;
        const scores = Array(n).fill().map(() => Array(n).fill(0));
        for (let i = 0; i < n; i++) {
            for (let j = 0; j < n; j++) {
                if (i === j) continue;
                scores[i][j] = shapes[i].color === shapes[j].color ? 5.0 : 0.5;
            }
        }
        return scores;
    }

    _alignment_scores(shapes) {
        const n = shapes.length;
        const scores = Array(n).fill().map(() => Array(n).fill(0));
        for (let i = 0; i < n; i++) {
            const [cx_i, cy_i] = this._center(shapes[i]);
            for (let j = 0; j < n; j++) {
                if (i === j) continue;
                const [cx_j, cy_j] = this._center(shapes[j]);
                const x_align = Math.max(0, 2 - Math.abs(cx_i - cx_j));
                const y_align = Math.max(0, 2 - Math.abs(cy_i - cy_j));
                scores[i][j] = x_align + y_align;
            }
        }
        return scores;
    }

    _containment_scores(shapes) {
        const n = shapes.length;
        const scores = Array(n).fill().map(() => Array(n).fill(0));
        for (let i = 0; i < n; i++) {
            for (let j = 0; j < n; j++) {
                if (i === j) continue;
                const [cx_j, cy_j] = this._center(shapes[j]);
                if (cx_j >= shapes[i].x && cx_j <= shapes[i].x + shapes[i].w &&
                    cy_j >= shapes[i].y && cy_j <= shapes[i].y + shapes[i].h) {
                    scores[i][j] = 5.0;
                } else {
                    scores[i][j] = 0.2;
                }
            }
        }
        return scores;
    }
}

File: keynote/shapes_core.py
"""
🛸 Shapes Core: Zara's Tools for Understanding Shape World
============================================================

All four demo engines in one file:
  1. ShapeTokenizer  — decompose scenes into parts (pixel / shape / part-level)
  2. ShapeEmbedding  — map shapes into a meaning space where similar = nearby
  3. ShapeVectorDB   — searchable library of shape patterns
  4. ShapeAttention   — multi-head attention showing which shapes relate

These power the live demo moments in the keynote talk.

Usage:
    from shapes_core import ShapeTokenizer, ShapeEmbedding, ShapeVectorDB, ShapeAttention
"""

import numpy as np
from typing import Dict, List, Tuple, Optional, Any
from collections import Counter, defaultdict

from shape_data import (
    SHAPE_TYPES, SHAPE_TYPE_TO_ID, ID_TO_SHAPE_TYPE,
    SHAPE_PARTS, PART_TO_ID, SHAPE_DECOMPOSITION,
    SAMPLE_SCENES, LAYER_COLORS, scene_to_feature_vector,
)


# =========================================================================
# PART 1: SHAPE TOKENIZER — "What Am I Looking At?"
# =========================================================================

class ShapeTokenizer:
    """
    Zara's first tool: breaking a visual scene into manageable pieces.

    Three strategies mirror text tokenization:
      - 'pixel':   rasterize to a grid (like char-level — huge, complete)
      - 'shape':   catalog whole shapes (like word-level — compact, brittle)
      - 'part':    decompose into sub-shape parts (like BPE — smart, flexible)

    Example:
        tok = ShapeTokenizer(strategy='part')
        tok.train(list(SAMPLE_SCENES.values()))
        tokens = tok.encode(SAMPLE_SCENES['house'])
        # [0, 3, 0, 3, 0, 3, 0, 3, 5, ...]  (part IDs)
    """

    def __init__(self, strategy: str = 'part'):
        if strategy not in ('pixel', 'shape', 'part'):
            raise ValueError(f"Strategy must be 'pixel', 'shape', or 'part', got '{strategy}'")
        self.strategy = strategy
        self.vocab: Dict[str, int] = {}
        self.inverse_vocab: Dict[int, str] = {}
        self.is_trained = False

    def train(self, scenes: List[Dict[str, Any]]) -> None:
        """Build vocabulary from a list of scenes."""
        if self.strategy == 'pixel':
            self._train_pixel()
        elif self.strategy == 'shape':
            self._train_shape(scenes)
        else:
            self._train_part()
        self.is_trained = True

    def _train_pixel(self) -> None:
        """Pixel strategy: vocabulary is a finite set of color values."""
        colors = ["empty"] + list(LAYER_COLORS.keys())
        self.vocab = {c: i for i, c in enumerate(colors)}
        self.inverse_vocab = {i: c for c, i in self.vocab.items()}

    def _train_shape(self, scenes: List[Dict[str, Any]]) -> None:
        """Shape strategy: every unique shape-label becomes a token."""
        self.vocab = {"<UNK>": 0}
        idx = 1
        for scene in scenes:
            for s in scene.get("shapes", []):
                key = f'{s["type"]}_{s["color"]}'
                if key not in self.vocab:
                    self.vocab[key] = idx
                    idx += 1
        self.inverse_vocab = {i: t for t, i in self.vocab.items()}

    def _train_part(self) -> None:
        """Part strategy: vocabulary is the set of sub-shape primitives."""
        self.vocab = {p: i for i, p in enumerate(SHAPE_PARTS)}
        self.inverse_vocab = {i: p for p, i in self.vocab.items()}

    def encode(self, scene: Dict[str, Any]) -> List[int]:
        """Encode a scene into a list of token IDs."""
        if not self.is_trained:
            raise RuntimeError("Tokenizer must be trained before encoding")

        if self.strategy == 'pixel':
            return self._encode_pixel(scene)
        elif self.strategy == 'shape':
            return self._encode_shape(scene)
        else:
            return self._encode_part(scene)

    def _encode_pixel(self, scene: Dict[str, Any], grid_size: int = 10) -> List[int]:
        """Rasterize scene to a grid. Each cell gets the color of whatever shape covers it."""
        grid = [[0] * grid_size for _ in range(grid_size)]
        empty_id = self.vocab["empty"]
        for s in scene.get("shapes", []):
            color_id = self.vocab.get(s["color"], empty_id)
            x0, y0 = int(s["x"]), int(s["y"])
            x1 = min(x0 + int(np.ceil(s["w"])), grid_size)
            y1 = min(y0 + int(np.ceil(s["h"])), grid_size)
            for r in range(max(y0, 0), y1):
                for c in range(max(x0, 0), x1):
                    grid[r][c] = color_id
        return [cell for row in grid for cell in row]

    def _encode_shape(self, scene: Dict[str, Any]) -> List[int]:
        """One token per shape."""
        unk = self.vocab.get("<UNK>", 0)
        tokens = []
        for s in scene.get("shapes", []):
            key = f'{s["type"]}_{s["color"]}'
            tokens.append(self.vocab.get(key, unk))
        return tokens

    def _encode_part(self, scene: Dict[str, Any]) -> List[int]:
        """Decompose each shape into sub-parts."""
        tokens = []
        for s in scene.get("shapes", []):
            parts = SHAPE_DECOMPOSITION.get(s["type"], ["straight_edge"])
            for p in parts:
                tokens.append(self.vocab.get(p, 0))
        return tokens

    def decode(self, token_ids: List[int]) -> List[str]:
        """Convert token IDs back to human-readable names."""
        return [self.inverse_vocab.get(tid, "?") for tid in token_ids]

    @property
    def vocab_size(self) -> int:
        return len(self.vocab)


# =========================================================================
# PART 2: SHAPE EMBEDDING — "The Meaning Map"
# =========================================================================

class ShapeEmbedding:
    """
    Zara's meaning map: shapes that appear together get similar vectors.

    Three strategies:
      - 'random':         random vectors (baseline — no real meaning)
      - 'spatial':        co-occurrence based (shapes near each other get similar vectors)
      - 'feature':        feature-engineered vectors from scene properties

    Example:
        emb = ShapeEmbedding(strategy='spatial', dimensions=16)
        emb.train(list(SAMPLE_SCENES.values()))
        emb.most_similar('house')  # [('car', 0.82), ('boat', 0.71), ...]
    """

    def __init__(self, strategy: str = 'spatial', dimensions: int = 16):
        if strategy not in ('random', 'spatial', 'feature'):
            raise ValueError(f"Strategy must be 'random', 'spatial', or 'feature', got '{strategy}'")
        self.strategy = strategy
        self.dimensions = dimensions
        self.scene_names: List[str] = []
        self.embeddings: Optional[np.ndarray] = None
        self.is_trained = False

    def train(self, scenes: Dict[str, Dict[str, Any]]) -> None:
        """
        Train embeddings from a dict of named scenes.

        Args:
            scenes: dict mapping scene name -> scene dict (like SAMPLE_SCENES)
        """
        self.scene_names = list(scenes.keys())
        n = len(self.scene_names)

        if self.strategy == 'random':
            np.random.seed(42)
            self.embeddings = np.random.randn(n, self.dimensions).astype(np.float32)
        elif self.strategy == 'feature':
            self.embeddings = np.array([
                scene_to_feature_vector(scenes[name], self.dimensions)
                for name in self.scene_names
            ], dtype=np.float32)
        else:  # spatial
            self._train_spatial(scenes)

        # Normalize rows
        norms = np.linalg.norm(self.embeddings, axis=1, keepdims=True)
        norms[norms == 0] = 1.0
        self.embeddings = self.embeddings / norms
        self.is_trained = True

    def _train_spatial(self, scenes: Dict[str, Dict[str, Any]]) -> None:
        """
        Build co-occurrence-style embeddings: scenes that share shape types
        and color patterns get similar vectors.
        """
        n = len(self.scene_names)
        # Build a feature matrix from shape-type and color distributions
        raw = np.zeros((n, len(SHAPE_TYPES) + len(LAYER_COLORS) + 4), dtype=np.float32)
        for i, name in enumerate(self.scene_names):
            shapes = scenes[name]["shapes"]
            for s in shapes:
                sid = SHAPE_TYPE_TO_ID.get(s["type"], 0)
                raw[i, sid] += 1
                cidx = list(LAYER_COLORS.keys()).index(s["color"]) if s["color"] in LAYER_COLORS else 0
                raw[i, len(SHAPE_TYPES) + cidx] += 1
            # Add spatial features
            if shapes:
                xs = [s["x"] + s["w"] / 2 for s in shapes]
                ys = [s["y"] + s["h"] / 2 for s in shapes]
                raw[i, -4] = np.mean(xs) / 10
                raw[i, -3] = np.mean(ys) / 10
                raw[i, -2] = len(shapes) / 10
                raw[i, -1] = len(set(s["type"] for s in shapes)) / len(SHAPE_TYPES)

        # Reduce to target dimensions via random projection (lightweight PCA alternative)
        np.random.seed(42)
        d = raw.shape[1]
        proj = np.random.randn(d, self.dimensions).astype(np.float32) / np.sqrt(d)
        self.embeddings = raw @ proj

    def get_embedding(self, name: str) -> np.ndarray:
        """Get the embedding vector for a named scene."""
        idx = self.scene_names.index(name)
        return self.embeddings[idx]

    def most_similar(self, name: str, top_k: int = 5) -> List[Tuple[str, float]]:
        """Find the most similar scenes to a given one."""
        vec = self.get_embedding(name)
        scores = self.embeddings @ vec  # cosine similarity (already normalized)
        ranked = np.argsort(-scores)
        results = []
        for idx in ranked:
            scene_name = self.scene_names[idx]
            if scene_name != name:
                results.append((scene_name, float(scores[idx])))
            if len(results) >= top_k:
                break
        return results

    def get_2d_projection(self) -> np.ndarray:
        """Project embeddings to 2D for visualization (simple PCA)."""
        centered = self.embeddings - self.embeddings.mean(axis=0)
        # SVD-based 2D projection
        U, S, Vt = np.linalg.svd(centered, full_matrices=False)
        return centered @ Vt[:2].T


# =========================================================================
# PART 3: SHAPE VECTOR DB — "The Shape Library"
# =========================================================================

class ShapeVectorDB:
    """
    Zara's searchable pattern library.

    Stores scene embeddings and supports similarity search.

    Example:
        db = ShapeVectorDB(dimensions=16)
        db.add('house', house_vector)
        results = db.search(query_vector, top_k=3)
        # [('house', 0.91), ('car', 0.78), ...]
    """

    def __init__(self, dimensions: int = 16):
        self.dimensions = dimensions
        self.vectors: Dict[str, np.ndarray] = {}
        self.id_list: List[str] = []

    def add(self, name: str, vector: np.ndarray) -> None:
        """Add a pattern to the library."""
        norm = np.linalg.norm(vector)
        if norm > 0:
            vector = vector / norm
        self.vectors[name] = vector.copy()
        if name not in self.id_list:
            self.id_list.append(name)

    def add_batch(self, names: List[str], vectors: np.ndarray) -> None:
        """Add multiple patterns at once."""
        for name, vec in zip(names, vectors):
            self.add(name, vec)

    def search(self, query: np.ndarray, top_k: int = 5) -> List[Tuple[str, float]]:
        """Find the most similar patterns to a query vector."""
        if not self.id_list:
            return []
        norm = np.linalg.norm(query)
        if norm > 0:
            query = query / norm
        mat = np.array([self.vectors[n] for n in self.id_list])
        scores = mat @ query
        ranked = np.argsort(-scores)
        results = []
        for idx in ranked[:top_k]:
            results.append((self.id_list[idx], float(scores[idx])))
        return results

    @property
    def size(self) -> int:
        return len(self.id_list)


# =========================================================================
# PART 4: SHAPE ATTENTION — "The Spotlight"
# =========================================================================

def softmax(x: np.ndarray, axis: int = -1) -> np.ndarray:
    """Softmax: turn scores into probabilities."""
    exp_x = np.exp(x - np.max(x, axis=axis, keepdims=True))
    return exp_x / np.sum(exp_x, axis=axis, keepdims=True)


class ShapeAttention:
    """
    Zara's spotlight system: which shapes should pay attention to each other?

    Four attention heads, each looking for a different relationship:
      - Head 0 (Proximity):    shapes that are physically close
      - Head 1 (Color):        shapes that share a color
      - Head 2 (Alignment):    shapes that are horizontally/vertically aligned
      - Head 3 (Containment):  shapes where one is inside another

    Example:
        att = ShapeAttention()
        weights = att.compute_attention(SAMPLE_SCENES['house'])
        # weights shape: (4, num_shapes, num_shapes)
    """

    HEAD_NAMES = ["Proximity", "Color Match", "Alignment", "Containment"]

    def __init__(self):
        pass

    def compute_attention(self, scene: Dict[str, Any]) -> Tuple[np.ndarray, List[str]]:
        """
        Compute multi-head attention weights for all shapes in a scene.

        Returns:
            weights: array of shape (4, n, n) — attention weights per head
            labels:  list of shape labels
        """
        shapes = scene["shapes"]
        n = len(shapes)
        labels = [s["label"] for s in shapes]

        if n == 0:
            return np.zeros((4, 0, 0)), []

        # Compute raw score matrices per head
        proximity = self._proximity_scores(shapes)
        color = self._color_scores(shapes)
        alignment = self._alignment_scores(shapes)
        containment = self._containment_scores(shapes)

        # Stack and apply softmax per row to get attention distributions
        raw = np.stack([proximity, color, alignment, containment])  # (4, n, n)
        weights = np.zeros_like(raw)
        for h in range(4):
            weights[h] = softmax(raw[h], axis=-1)

        return weights, labels

    def _center(self, s: Dict) -> Tuple[float, float]:
        return s["x"] + s["w"] / 2, s["y"] + s["h"] / 2

    def _proximity_scores(self, shapes: List[Dict]) -> np.ndarray:
        """Closer shapes get higher scores."""
        n = len(shapes)
        scores = np.zeros((n, n))
        for i in range(n):
            cx_i, cy_i = self._center(shapes[i])
            for j in range(n):
                if i == j:
                    continue
                cx_j, cy_j = self._center(shapes[j])
                dist = np.sqrt((cx_i - cx_j) ** 2 + (cy_i - cy_j) ** 2)
                scores[i, j] = max(0, 10 - dist)  # inverse distance, clamped
        return scores

    def _color_scores(self, shapes: List[Dict]) -> np.ndarray:
        """Same-color shapes get high scores."""
        n = len(shapes)
        scores = np.zeros((n, n))
        for i in range(n):
            for j in range(n):
                if i == j:
                    continue
                if shapes[i]["color"] == shapes[j]["color"]:
                    scores[i, j] = 5.0
                else:
                    scores[i, j] = 0.5
        return scores

    def _alignment_scores(self, shapes: List[Dict]) -> np.ndarray:
        """Shapes that share x-center or y-center get high scores."""
        n = len(shapes)
        scores = np.zeros((n, n))
        for i in range(n):
            cx_i, cy_i = self._center(shapes[i])
            for j in range(n):
                if i == j:
                    continue
                cx_j, cy_j = self._center(shapes[j])
                x_align = max(0, 2 - abs(cx_i - cx_j))
                y_align = max(0, 2 - abs(cy_i - cy_j))
                scores[i, j] = x_align + y_align
        return scores

    def _containment_scores(self, shapes: List[Dict]) -> np.ndarray:
        """Shape j inside shape i gets high score."""
        n = len(shapes)
        scores = np.zeros((n, n))
        for i in range(n):
            for j in range(n):
                if i == j:
                    continue
                # Check if center of j is inside bounding box of i
                cx_j, cy_j = self._center(shapes[j])
                if (shapes[i]["x"] <= cx_j <= shapes[i]["x"] + shapes[i]["w"] and
                        shapes[i]["y"] <= cy_j <= shapes[i]["y"] + shapes[i]["h"]):
                    scores[i, j] = 5.0
                else:
                    scores[i, j] = 0.2
        return scores

    def combined_attention(self, scene: Dict[str, Any]) -> Tuple[np.ndarray, List[str]]:
        """Average attention across all heads."""
        weights, labels = self.compute_attention(scene)
        if weights.size == 0:
            return np.zeros((0, 0)), labels
        combined = weights.mean(axis=0)
        return combined, labels
